File: javascript/kruskal.js
class DisjointSet {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = Array(n).fill(0);
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX !== rootY) {
            if (this.rank[rootX] > this.rank[rootY]) {
                this.parent[rootY] = rootX;
            } else if (this.rank[rootX] < this.rank[rootY]) {
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX;
                this.rank[rootX]++;
            }
        }
    }
}

class Graph {
    constructor(vertices) {
        this.V = vertices;
        this.edges = [];
    }

    addEdge(u, v, w) {
        this.edges.push([u, v, w]);
    }

    kruskalMST() {
        this.edges.sort((a, b) => a[2] - b[2]);
        const disjointSet = new DisjointSet(this.V);
        const result = [];

        for (const [u, v, w] of this.edges) {
            const setU = disjointSet.find(u);
            const setV = disjointSet.find(v);

            if (setU !== setV) {
                result.push([u, v, w]);
                disjointSet.union(setU, setV);
            }
        }

        return result;
    }
}

// Example usage:
const graph = new Graph(4);
graph.addEdge(0, 1, 10);
graph.addEdge(0, 2, 6);
graph.addEdge(0, 3, 5);
graph.addEdge(1, 3, 15);
graph.addEdge(2, 3, 4);

const mst = graph.kruskalMST();
console.log("Edges in the Minimum Spanning Tree:");
mst.forEach(([u, v, w]) => {
    console.log(`(${u}, ${v}) -> ${w}`);
});

File: ruby/kruskal.rb
class DisjointSet
  def initialize(n)
    @parent = Array.new(n) { |i| i }
    @rank = Array.new(n, 0)
  end

  def find(x)
    @parent[x] = find(@parent[x]) if @parent[x] != x
    @parent[x]
  end

  def union(x, y)
    root_x = find(x)
    root_y = find(y)

    if root_x != root_y
      if @rank[root_x] > @rank[root_y]
        @parent[root_y] = root_x
      elsif @rank[root_x] < @rank[root_y]
        @parent[root_x] = root_y
      else
        @parent[root_y] = root_x
        @rank[root_x] += 1
      end
    end
  end
end

class Graph
  def initialize(vertices)
    @V = vertices
    @edges = []
  end

  def add_edge(u, v, w)
    @edges << [u, v, w]
  end

  def kruskal_mst
    @edges.sort_by! { |edge| edge[2] }
    disjoint_set = DisjointSet.new(@V)
    result = []

    @edges.each do |u, v, w|
      set_u = disjoint_set.find(u)
      set_v = disjoint_set.find(v)

      if set_u != set_v
        result << [u, v, w]
        disjoint_set.union(set_u, set_v)
      end
    end

    result
  end
end

# Example usage:
graph = Graph.new(4)
graph.add_edge(0, 1, 10)
graph.add_edge(0, 2, 6)
graph.add_edge(0, 3, 5)
graph.add_edge(1, 3, 15)
graph.add_edge(2, 3, 4)

mst = graph.kruskal_mst
puts "Edges in the Minimum Spanning Tree:"
mst.each do |u, v, w|
  puts "(#{u}, #{v}) -> #{w}"
end
